### ISSUE TITLE
Type assertion suddenly required for a default-typed generic type when directly assigned to an optional object property

### ISSUE THREAD

**Issue by joelposti** [microsoft/TypeScript#44084]
# Bug Report

It was discovered that the return value of a function cannot be directly assigned to an optional object property without doing a type assertion, if the function returns a default-typed generic type. This behaviour is unexpected because the type assertion is not needed when assigning to a required object property. Neither is the type assertion needed when the return value is first assigned to a constant and then the constant is assigned to the optional object property.

To reproduce this behaviour tsc needs to be run with the `--strictNullChecks` option.

### 🔎 Search Terms

optional parameter, optional object property, object property with default value, parameter default value, generic default type

### 🕗 Version & Regression Information

The issue was discovered when using typescript@4.0.3. It's likely present in earlier versions as well. The issue is reproducible with typescript@4.2.4 and typescript@4.4.0-dev.20210514 (current typescript@next).

I read https://github.com/Microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs and https://www.typescriptlang.org/tsconfig#strictNullChecks carefully. I could not figure out whether the reported behaviour is a bug or working as intended.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEAkFMCdIcgM6gIagGYFcB2BjALgJYD2WoeAFsnqLHhtFkqgG7IA2GkoRao2A1liIB3UngCeABy5poRALYpQCIvRxcKVAFAhQOZFiHUARpGAEkR5XioF9bNuLJTIAEwB0EEZGYwANHwIGOyOZORcrBxcFkoIeNAEWADmngAqRDpg8sj8GuHo2PjEpJDICAQwZESB0dQxEtKuVTSQdAygSZBYMHbO0qCpYdq6rpBoyBhseEh41XEJye5aDVwACkQI5cZskABKrfRYqS5IALzWC0mgAD6gWBjyptBamLiEJB2tAGrsnADyaAAqlhBCIji4ADyDSAADzwXVcSHWmwI2z2BwYx2kZwuiSSAD4ABQASgAXANQABvLSgFptUhwZDGHBwFBIVJaAC+WkyoAA6kRoPwEEscCQ4ihyfM8aBzp08D8ogDgaDRFjICSeYl4dBxupQPsAI4YAiwVyrOTSaASKk00CSZDQZDyACMUvieK5PN0AqFIq0YsYJnJRpNZotRCtNvO1NpDqdrvJ8sV-yBIKEapcmu5Wm1MD1XD+kneWHYEajTlj9sdzpdAH53ZcvbyACJESCWIjUYSC-ieACSCCwcGoFC4o3Gk2otEOfS4PE+Ct+kGV6bB6t5qGlyTrAfF1Bw5KLJbLlpg0dtcZricXKZXadV4Ok2e9YF9ws8fPIThiAAMIduBK-nckDqJsjriLugYSq4R7FsUp6RueTgxna8a1km3zLquj7qgBHrJESxLNj6vYivy36gBYw6ji4bIIOexQgW4bigOEsBQfuoCRJwsq3thD4Zk+GrEdB1CQHBJ5sOWyF8VW6E3jxkBckAA)

### 💻 Code

```ts
// Here's a function that returns a value of unknown type from a source that
// cannot be/is not statically typed. However, usually the value is a string. To
// make the function easier to use it is typed to return generic type T that
// defaults to string.
type PossibleReturnTypes = string | number
function getValueOfUnknownType<T extends PossibleReturnTypes = string>(): T {
  return 'abc' as T
}

// Works.
const a: string = getValueOfUnknownType()

interface RequiredProperty {
  param1: string
}

// Works.
const b: RequiredProperty = {
  param1: getValueOfUnknownType()
}

interface OptionalProperty {
  param1?: string
}

// Does not work. Isn't the default return type of getValueOfUnknownType
// a string?
const c: OptionalProperty = {
  param1: getValueOfUnknownType()
}

// Works. Why is `<string>` necessary?
const d: OptionalProperty = {
  param1: getValueOfUnknownType<string>()
}

// Works. Why isn't type assertion needed here?
const value = getValueOfUnknownType()
const e: OptionalProperty = {
  param1: value
}
```

### 🙁 Actual behavior

Tsc reports an error on line 29:

```
Type 'PossibleReturnTypes' is not assignable to type 'string | undefined'.
  Type 'number' is not assignable to type 'string | undefined'.
```

### 🙂 Expected behavior

No error since the default return type of getValueOfUnknownType is a string.

**Comment by RyanCavanaugh** [comment]
This is a really bad API pattern and putting type parameters in that position will lead to all sorts of trouble. It's best not to do this and you will encounter more problems down the line.

Anyway what's happening is that we have logic to make this API pattern slightly less bad which says that a last-ditch inference candidate can be sourced from the contextual type if it's assignable to the type parameter's constraint. In the error case the contextual type is `string | undefined`, which isn't assignable to `string | number`, so it's rejected as a candidate and we fall back to the constraint since there were zero matching candidates but nonzero overall candidates, thus `T = string | number`. Each step in that logic is justifiable and enables other scenarios, so the fact that this one fails to work as expected is just a limitation of using the pattern in the first place.

**Comment by joelposti** [comment]
Thank you for your comment!

> This is a really bad API pattern and putting type parameters in that position will lead to all sorts of trouble. It's best not to do this and you will encounter more problems down the line.

Could you help illustrate what makes the API pattern so bad? I'm aware of the inherent dangers of this approach as getValueOfUnknownType might not return a string (as explained in the comment in the code), if that is what you mean. Rest assured, there are safety mechanisms in the real life version of getValueOfUnknownType which I did not include here in order to keep the example code short and to the point. A solution like it will only serve as a temporary fix until a proper type safe implementation can be done.

But is there something else that is bad that I'm oblivous to? I wish to learn and improve.

> Anyway what's happening is that we have logic to make this API pattern slightly less bad which says that a last-ditch inference candidate can be sourced from the contextual type if it's assignable to the type parameter's constraint. In the error case the contextual type is string | undefined, which isn't assignable to string | number, so it's rejected as a candidate and we fall back to the constraint since there were zero matching candidates but nonzero overall candidates, thus T = string | number. Each step in that logic is justifiable and enables other scenarios, so the fact that this one fails to work as expected is just a limitation of using the pattern in the first place.

Okay. If the contextual type is `string | undefined` in the `const c` case, is the contextual type something different in the `const e` case? What makes `const value` fit it into `param1?: string` when `getValueOfUnknownType()` does not?

**Comment by RyanCavanaugh** [comment]
> Okay. If the contextual type is string | undefined in the const c case, is the contextual type something different in the const e case?

Yep, contextual typing can't "flow backwards" into prior statements. When processing this call
```ts
const value = getValueOfUnknownType()
const e: OptionalProperty = {
  param1: value
}
```
the existence of `e` isn't considered at all, so there is no contextual type present, so `value` gets bound to `string` via the default.

**Comment by joelposti** [comment]
Thank you for taking your time to answer my questions. Much appreciated.

I got no further comments regarding the behaviour I've reported other than that I understand the technical reason for it, but from TypeScript user's perspective the difference between `const c` and `const e` cases is strange. I leave it to you and other TypeScript developers to decide whether it is something that needs fixing or not.
